### PR TITLE
issue-792: Forbid Union[T, None]

### LINKED ISSUE #792: Forbid `Union[T, None]`, use `Optional[T]` instead
# Rule request

## Thesis

One can write `Optional[T]` to indicate that `T` and `None` can be returned.
This is a valid way to do it.

But `Union[T, None]` is invalid. And should be forbidden. 

I have to note that `Union[A, B, C, None]` is ok.

## Reasoning

This is a consistency rule.

### PR DESCRIPTION
**Pull request by novikovfred**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

- Closes #792

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ We used to have incremental versioning before `0.1.0`.
 - Forbids using nested `typing.Literal`, `typing.Union` and `typing.Annotated`
 - Forbids use of vague import names (e.g. `from json import loads`)
 - Make ``OveruseOfNoqaCommentViolation`` configurable (`--max-noqa-comments`)
+- Forbid `Union[T, None]` in function annotations
```
Comment (sobolevn) [inline CHANGELOG.md]: `in annotations`

File: tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py [hunk 2]
```diff
@@ -130,6 +147,14 @@ def function(arg: typing.Literal[typing.Literal[typing.Literal[1]]]): ...
 def function(arg: typing.Union[Union[int], str]): ...
 """
 
+wrong_arg_union_with_none_annotation = """
+def function(empty_arg: Union[int, None]): ...
+"""
+
+wrong_embedded_arg_union_with_none_annotation = """
+def function(empty_arg: Union[Union[int, None], Optional[int]]): ...
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py]: Please, also cover class attributes and variable annotations.

File: tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py [hunk 3]
```diff
@@ -227,6 +252,10 @@ def test_wrong_argument_annotation(
     correct_multiline_arguments,
     correct_arg_none_annotation,
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py]: I guess that I have missed this point with nested `Literal`, `Union`, and `Annotated` types. Can you please also cover it here as well. I know that's not 100% related. But, it's pretty close.

File: tests/test_visitors/test_ast/test_annotations/test_return_annotations.py [hunk 4]
```diff
@@ -58,6 +59,22 @@ def function() -> typing.Literal[1, 2, 3]: ...
 def function() -> Union[typing.Literal[2, 3], int]: ...
 """
 
+correct_union_with_none_return = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_annotations/test_return_annotations.py]: The same about class-level annotation and variables goes here.

File: tests/test_visitors/test_ast/test_annotations/test_return_annotations.py [hunk 5]
```diff
@@ -58,6 +59,22 @@ def function() -> typing.Literal[1, 2, 3]: ...
...
+def function() -> Union[int, float, None]: ...
+"""
+
+correct_embedded_union_with_none_return = """
+def function() -> Union[Union[int, float, None], float]: ...
+"""
+
+correct_union_return = """
+def function() -> Union[True, float]: ...
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_annotations/test_return_annotations.py]: That's not correct. `True` it not a valid type. Let's use `Union[bool, float]`

File: wemake_python_styleguide/violations/annotations.py [hunk 6]
```diff
@@ -91,3 +93,34 @@ class NestedAnnotationsViolation(ASTViolation):
...
     code = 702
+
+
+@final
+class UnionNoneViolation(ASTViolation):
+    """
+    Forbids to use ``Union[T, None]`` typing annotation.
+
+    Reasoning:
+        Union[T, None] is just the same as Optional[T].
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/annotations.py]: `Optional[T]` is not in double backquotes.

File: wemake_python_styleguide/violations/annotations.py [hunk 7]
```diff
@@ -91,3 +93,34 @@ class NestedAnnotationsViolation(ASTViolation):
...
+        There's no need to use the first version.
+        It is not type related, it is a consistency rule.
+
+    Solution:
+        Replace ``Union[T, None]`` with ``Optional[T]``.
+
+    Example::
+
+        # Correct:
+        def func(empty: Optional[T]):
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/annotations.py]: Let's use variables as examples. They take less space. And no problem with `'''`

File: tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py [hunk 8]
```diff
@@ -67,6 +68,22 @@ def function(arg: typing.Literal[1, 2, 3]): ...
 def function(arg: typing.Union[Literal[1]]): ...
 """
 
+correct_arg_union_with_none_annotation = """
+def function(empty_arg: Union[int, float, None]): ...
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py]: Can you please also test `typing.Union`?

File: wemake_python_styleguide/visitors/ast/annotations.py [hunk 9]
```diff
@@ -124,23 +126,46 @@ def visit_arg(self, node: ast.arg) -> None:
...
 
         if annotation_name.id != 'Literal':
             return
 
-        if not isinstance(slice_index.value, ast.NameConstant):
+        if self._check_slice_arg_for_none(slice_index.value):
+            self.add_violation(LiteralNoneViolation(node))
+
+    def _check_for_union_none(self, node: ast.Subscript) -> None:
+        annotation_name = cast(ast.Name, node.value)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/annotations.py]: Can you please remove `cast`? It is not recommended use `cast`.

File: tests/test_visitors/test_ast/test_annotations/test_return_annotations.py (hunk 5 above)
Comment (novikovfred) [reply tests/test_visitors/test_ast/test_annotations/test_return_annotations.py]: I just want to test if statement: `if isinstance(slice_arg, ast.NameConstant):` will it work with bool?

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_annotations/test_return_annotations.py]: Ok, I see. Then it is fine.

File: tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py (hunk 3 above)
Comment (novikovfred) [reply tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py]: How I understand I need to cover `class attributes and variable annotations.`
Now, `visitor` checks only the functions. So I need some more time to do it.

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py]: Sure, no hurry.

**Comment by sobolevn** [comment]
Hi, @novikovfred! How's it going?

**Comment by novikovfred** [comment]
Working on this problem. I think, i’ll finish this weekend.

File: wemake_python_styleguide/visitors/ast/annotations.py (hunk 9 above)
Comment (novikovfred) [reply wemake_python_styleguide/visitors/ast/annotations.py]: If I remove `cast` expressions, I'll have some problems with mypy.

**Comment by novikovfred** [comment]
Ok, I'll refactor due to new logic.

**Pull request closed without merging**